File: solutions/prospective/algorithms/cross_lang/quicksort_hoare.c
/*
Quicksort with Hoare Partition Scheme

Time Complexity: O(n log n) avg
Space Complexity: O(log n) stack

Hoare's partition is often more efficient than Lomuto's (fewer swaps).
*/
=== MEMO START ===
int partition(int arr[], int low, int high) {
    int pivot = arr[low];
    int i = low - 1;
    int j = high + 1;
    
    while (1) {
        do {
            i++;
        } while (arr[i] < pivot);
        
        do {
            j--;
        } while (arr[j] > pivot);
        
        if (i >= j)
            return j;
            
        // Swap arr[i] and arr[j]
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

void quick_sort(int arr[], int low, int high) {
    if (low < high) {
        int p = partition(arr, low, high);
        quick_sort(arr, low, p);
        quick_sort(arr, p + 1, high);
    }
}

File: solutions/focus/quick_sort_hoare.py
"""
Quicksort (Hoare Partition)

Time Complexity: O(n log n) avg
Space Complexity: O(log n) stack

Hoare's partition is generally more efficient than Lomuto's (fewer swaps).
"""
=== MEMO START ===
def partition(arr, low, high):
    pivot = arr[low]
    i = low - 1
    j = high + 1
    
    while True:
        i += 1
        while arr[i] < pivot:
            i += 1
        j -= 1
        while arr[j] > pivot:
            j -= 1  
        if i >= j:
            return j
        arr[i], arr[j] = arr[j], arr[i]

def quick_sort(arr, low, high):
    if low < high:
        p = partition(arr, low, high)
        quick_sort(arr, low, p)
        quick_sort(arr, p + 1, high)